core/src/main/java/org/apache/calcite/adapter/jdbc/JdbcSchema.java

```diff
@@ -237,80 +252,86 @@ public DataSource getDataSource() {
+        getTableType(tableDef.tableType));
   }
 
-  private ImmutableMap<String, JdbcTable> computeTables() {
-    Connection connection = null;
-    ResultSet resultSet = null;
+  private static MetaImpl.MetaTable metaDataMapper(ResultSet resultSet) {
     try {
-      connection = dataSource.getConnection();
-      final Pair<@Nullable String, @Nullable String> catalogSchema = getCatalogSchema(connection);
-      final String catalog = catalogSchema.left;
-      final String schema = catalogSchema.right;
-      final Iterable<MetaImpl.MetaTable> tableDefs;
-      Foo threadMetadata = THREAD_METADATA.get();
-      if (threadMetadata != null) {
-        tableDefs = threadMetadata.apply(catalog, schema);
-      } else {
-        final List<MetaImpl.MetaTable> tableDefList = new ArrayList<>();
-        final DatabaseMetaData metaData = connection.getMetaData();
-        resultSet = metaData.getTables(catalog, schema, null, null);
-        while (resultSet.next()) {
-          final String catalogName = resultSet.getString(1);
-          final String schemaName = resultSet.getString(2);
-          final String tableName = resultSet.getString(3);
-          final String tableTypeName = resultSet.getString(4);
-          tableDefList.add(
-              new MetaImpl.MetaTable(catalogName, schemaName, tableName,
-                  tableTypeName));
-        }
-        tableDefs = tableDefList;
-      }
+      return new MetaImpl.MetaTable(intern(resultSet.getString(1)), intern(resultSet.getString(2)),
+          intern(resultSet.getString(3)),
+          intern(resultSet.getString(4)));
+    } catch (SQLException e) {
+      throw new RuntimeException(e);
+    }
+  }
 
-      final ImmutableMap.Builder<String, JdbcTable> builder =
-          ImmutableMap.builder();
-      for (MetaImpl.MetaTable tableDef : tableDefs) {
-        // Clean up table type. In particular, this ensures that 'SYSTEM TABLE',
-        // returned by Phoenix among others, maps to TableType.SYSTEM_TABLE.
-        // We know enum constants are upper-case without spaces, so we can't
-        // make things worse.
-        //
-        // PostgreSQL returns tableTypeName==null for pg_toast* tables
-        // This can happen if you start JdbcSchema off a "public" PG schema
-        // The tables are not designed to be queried by users, however we do
-        // not filter them as we keep all the other table types.
-        final String tableTypeName2 =
-            tableDef.tableType == null
+  private static @Nullable String intern(@Nullable String string) {
+    if (string == null) {
+      return null;
+    }
+    return string.intern();
+  }
+
+  private static TableType getTableType(String tableTypeName) {
+    // Clean up table type. In particular, this ensures that 'SYSTEM TABLE',
+    // returned by Phoenix among others, maps to TableType.SYSTEM_TABLE.
+    // We know enum constants are upper-case without spaces, so we can't
+    // make things worse.
```

Comment:
worse than what?

Reply:
I didn't change this comment. I just moved it around. Therefore, I'm not able to tell you.